packages/typir/src/typir.ts

```diff
@@ -55,46 +51,37 @@ export class Typir {
         typeRelationships: TypeRelationshipCaching;
         domainElementInference: DomainElementInferenceCaching;
     };
+    graph: TypeGraph;
+    kinds: KindRegistry;
     operators: OperatorManager;
     printer: ProblemPrinter;
     validation: {
         collector: ValidationCollector;
         constraints: ValidationConstraints;
     };
+};
 
-    constructor() {
-        this.assignability = new DefaultTypeAssignability(this);
-        this.equality = new DefaultTypeEquality(this);
-        this.conversion = new DefaultTypeConversion(this);
-        this.subtype = new DefaultSubType(this);
-        this.caching = {
-            typeRelationships: new DefaultTypeRelationshipCaching(this),
-            domainElementInference: new DefaultDomainElementInferenceCaching(this), // cached inference results, intended to be used by multiple inferring instances, other services should use the central inference service nevertheless
-        };
-        this.inference = new DefaultTypeInferenceCollector(this);
-        this.operators = new DefaultOperatorManager(this);
-        this.printer = new DefaultTypeConflictPrinter(this);
-        this.validation = {
-            collector: new DefaultValidationCollector(this),
-            constraints: new DefaultValidationConstraints(this),
-        };
-    }
-
-    // manage kinds
-    registerKind(kind: Kind): void {
-        const key = kind.$name;
-        if (this.kinds.has(key)) {
-            if (this.kinds.get(key) === kind) {
-                // that is OK
-            } else {
-                throw new Error(`duplicate kind named '${key}'`);
-            }
-        } else {
-            this.kinds.set(key, kind);
-        }
+export const DefaultTypirServiceModule: Module<TypirServices> = {
+    assignability: (services) => new DefaultTypeAssignability(services),
+    equality: (services) => new DefaultTypeEquality(services),
+    conversion: (services) => new DefaultTypeConversion(services),
+    graph: () =>  new TypeGraph(),
+    subtype: (services) => new DefaultSubType(services),
+    inference: (services) => new DefaultTypeInferenceCollector(services),
+    caching: {
+        typeRelationships: (services) => new DefaultTypeRelationshipCaching(services),
+        domainElementInference: () => new DefaultDomainElementInferenceCaching()
+    },
+    operators: (services) => new DefaultOperatorManager(services),
+    kinds: () => new DefaultKindRegistry(),
+    printer: () => new DefaultTypeConflictPrinter(),
+    validation: {
+        collector: (services) => new DefaultValidationCollector(services),
+        constraints: (services) => new DefaultValidationConstraints(services),
     }
+};
 
-    getKind(type: string): Kind | undefined {
-        return this.kinds.get(type)!;
-    }
+export function createTypirServices() {
```

Comment:
I would like to see the actual return type `: TypirServices`. It is not required, but improves the readability for users who are not familiar with our DI.

Reply:
I changed it myself 🙂